### PR TITLE
[LB] Implement `resource/opentelekomcloud_lb_loadbalancer_v3`

### PR DESCRIPTION
**Pull request by lego963**
## Summary of the Pull Request
Implement `resource/opentelekomcloud_lb_loadbalancer_v3`
Refers to: #1486

## PR Checklist

* [x] Refers to: #1486
* [x] Tests added/passed.
* [x] Documentation updated.
* [x] Schema updated.
* [x] Release notes added.

## Acceptance Steps Performed

```
=== RUN   TestAccLBV3LoadBalancer_basic
=== PAUSE TestAccLBV3LoadBalancer_basic
=== CONT  TestAccLBV3LoadBalancer_basic
--- PASS: TestAccLBV3LoadBalancer_basic (118.56s)
PASS

Process finished with the exit code 0
```

### CODE REVIEW

File: opentelekomcloud/services/elb/v3/common.go [hunk 1]
```diff
@@ -1,6 +1,34 @@
...
+import (
+	"context"
+	"fmt"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	golangsdk "github.com/opentelekomcloud/gophertelekomcloud"
+	"github.com/opentelekomcloud/terraform-provider-opentelekomcloud/opentelekomcloud/common/cfg"
+)
+
+type ContextKey string
```
Comment (outcatcher) [inline opentelekomcloud/services/elb/v3/common.go]: There is already `common.ClientCtxKey` type

File: opentelekomcloud/services/elb/v3/common.go [hunk 2]
```diff
@@ -1,6 +1,34 @@
...
+	if client != nil {
+		return client.(*golangsdk.ServiceClient), nil
+	}
+	config := meta.(*cfg.Config)
+	client, err := config.ElbV3Client(config.GetRegion(d))
+	if err != nil {
+		return nil, fmt.Errorf(ErrCreateClient, err)
+	}
+	return client.(*golangsdk.ServiceClient), nil
+}
```
Comment (outcatcher) [inline opentelekomcloud/services/elb/v3/common.go]: Please use methods from `common` package

File: opentelekomcloud/services/elb/v3/resource_opentelekomcloud_lb_loadbalancer_v3.go [hunk 3]
```diff
@@ -0,0 +1,391 @@
...
+		Target:       []string{"ACTIVE"},
+		Refresh:      resourceElbV3LoadBalancerRefreshFunc(client, lb.ID),
+		Timeout:      d.Timeout(schema.TimeoutCreate),
+		Delay:        5 * time.Second,
+		PollInterval: 1 * time.Second,
+	}
+	_, err = stateConf.WaitForStateContext(ctx)
+	if err != nil {
+		return diag.FromErr(err)
+	}
```
Comment (outcatcher) [inline opentelekomcloud/services/elb/v3/resource_opentelekomcloud_lb_loadbalancer_v3.go]: This wait is not actually needed

File: opentelekomcloud/services/elb/v3/resource_opentelekomcloud_lb_loadbalancer_v3.go [hunk 4]
```diff
@@ -0,0 +1,391 @@
...
+		Pending:      []string{"PENDING_UPDATE", "PENDING_DELETE", "ACTIVE"},
+		Refresh:      resourceElbV3LoadBalancerRefreshFunc(client, d.Id()),
+		Timeout:      d.Timeout(schema.TimeoutDelete),
+		Delay:        5 * time.Second,
+		PollInterval: 1 * time.Second,
+	}
+	_, err = stateConf.WaitForStateContext(ctx)
+	if err != nil {
+		return diag.FromErr(err)
+	}
```
Comment (outcatcher) [inline opentelekomcloud/services/elb/v3/resource_opentelekomcloud_lb_loadbalancer_v3.go]: Same as previous wait

File: opentelekomcloud/services/elb/v3/resource_opentelekomcloud_lb_loadbalancer_v3.go [hunk 5]
```diff
@@ -0,0 +1,391 @@
...
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+			"availability_zones": {
+				Type:     schema.TypeSet,
+				Required: true,
+				ForceNew: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+				Set:      schema.HashString,
```
Comment (outcatcher) [inline opentelekomcloud/services/elb/v3/resource_opentelekomcloud_lb_loadbalancer_v3.go]: `Set` argument is not required here, it's set by default

File: opentelekomcloud/services/elb/v3/resource_opentelekomcloud_lb_loadbalancer_v3.go [hunk 6]
```diff
@@ -0,0 +1,391 @@
...
+			},
+			"subnet_id": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"network_ids": {
+				Type:     schema.TypeSet,
+				Required: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+				Set:      schema.HashString,
```
Comment (outcatcher) [inline opentelekomcloud/services/elb/v3/resource_opentelekomcloud_lb_loadbalancer_v3.go]: `Set` argument can be skipped

File: opentelekomcloud/services/elb/v3/resource_opentelekomcloud_lb_loadbalancer_v3.go [hunk 7]
```diff
@@ -0,0 +1,391 @@
...
+		CreateContext: resourceLoadBalancerV3Create,
+		ReadContext:   resourceLoadBalancerV3Read,
+		UpdateContext: resourceLoadBalancerV3Update,
+		DeleteContext: resourceLoadBalancerV3Delete,
+
+		Timeouts: &schema.ResourceTimeout{
+			Create: schema.DefaultTimeout(10 * time.Minute),
+			Update: schema.DefaultTimeout(10 * time.Minute),
+			Delete: schema.DefaultTimeout(10 * time.Minute),
+		},
```
Comment (outcatcher) [inline opentelekomcloud/services/elb/v3/resource_opentelekomcloud_lb_loadbalancer_v3.go]: Timeouts are not required here. There are no actual waits.

File: opentelekomcloud/services/elb/v3/resource_opentelekomcloud_lb_loadbalancer_v3.go [hunk 8]
```diff
@@ -0,0 +1,391 @@
...
+	_, err = stateConf.WaitForStateContext(ctx)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	if len(lb.PublicIps) > 0 {
+		config := meta.(*cfg.Config)
+		nwV2Client, err := config.NetworkingV2Client(config.GetRegion(d))
+		if err != nil {
+			diag.FromErr(err)
```
Comment (outcatcher) [inline opentelekomcloud/services/elb/v3/resource_opentelekomcloud_lb_loadbalancer_v3.go]: Missing `return`

**Review by outcatcher — CHANGES_REQUESTED** [review]

File: opentelekomcloud/services/elb/v3/resource_opentelekomcloud_lb_loadbalancer_v3.go [hunk 9]
```diff
@@ -0,0 +1,340 @@
...
+		return config.ElbV3Client(config.GetRegion(d))
+	})
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	lb, err := loadbalancers.Get(client, d.Id()).Extract()
+	if err != nil {
+		return diag.FromErr(err)
+	}
```
Comment (outcatcher) [inline opentelekomcloud/services/elb/v3/resource_opentelekomcloud_lb_loadbalancer_v3.go:337]: You can avoid this request storing public IP values in the resource data

File: opentelekomcloud/services/elb/v3/resource_opentelekomcloud_lb_loadbalancer_v3.go [hunk 10]
```diff
@@ -89,6 +89,14 @@ func ResourceLoadBalancerV3() *schema.Resource {
...
 				Elem: &schema.Resource{
 					Schema: map[string]*schema.Schema{
+						"public_ip_id": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"public_ip_address": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
```
Comment (outcatcher) [inline opentelekomcloud/services/elb/v3/resource_opentelekomcloud_lb_loadbalancer_v3.go]: **Optional:** Those fields can be just `id` and `address`, as the parent field is named `public_ip`

File: opentelekomcloud/services/elb/v3/resource_opentelekomcloud_lb_loadbalancer_v3.go [hunk 11]
```diff
@@ -224,6 +232,14 @@ func resourceLoadBalancerV3Read(ctx context.Context, d *schema.ResourceData, met
 
 	log.Printf("[DEBUG] Retrieved loadbalancer %s: %#v", d.Id(), lb)
 
+	publicIpInfo := make([]map[string]interface{}, len(lb.PublicIps))
+	for i, v := range lb.PublicIps {
+		publicIpInfo[i] = map[string]interface{}{
+			"public_ip_id":      v.PublicIpID,
+			"public_ip_address": v.PublicIpAddress,
+		}
```
Comment (outcatcher) [inline opentelekomcloud/services/elb/v3/resource_opentelekomcloud_lb_loadbalancer_v3.go]: You should set other fields from the schema, e.g:
```
		publicIpInfo[i] = map[string]interface{}{
			"public_ip_id":      v.PublicIpID,
			"public_ip_address": v.PublicIpAddress,
			"bandwidth_name": d.Get("public_ip.0.bandwidth_name").(string)
			...
		}
```

or init it with a value from the schema:
```
info := d.Get("public_ip.0").(map[string]interface{})
info["public_ip_id"] = v.publicIpID
...
publicIpInfo[i] = info
```

(or replace `0` with `i` if you want to support a list of public IPs

File: opentelekomcloud/services/elb/v3/resource_opentelekomcloud_lb_loadbalancer_v3.go [hunk 12]
```diff
@@ -326,12 +343,13 @@ func resourceLoadBalancerV3Delete(ctx context.Context, d *schema.ResourceData, m
 
 	if len(lb.PublicIps) > 0 {
 		config := meta.(*cfg.Config)
-		nwV1Client, err := config.NetworkingV1Client(config.GetRegion(d))
+		nwV1Client, err := config.NetworkingV2Client(config.GetRegion(d))
```
Comment (outcatcher) [inline opentelekomcloud/services/elb/v3/resource_opentelekomcloud_lb_loadbalancer_v3.go]: it's `nwV2Client` now :)

**Review by outcatcher — CHANGES_REQUESTED** [review]

**Review by outcatcher — DISMISSED** [review]

File: docs/resources/lb_loadbalancer_v3.md [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+* `router_id` - (Optional) ID of the router (or VPC) this LoadBalancer belongs to. Changing
+  this creates a new LoadBalancer.
+
+* `subnet_id` - (Optional) The ID of the subnet to which the LoadBalancer belongs.
+
+-> `router_id` and `subnet_id` cannot be left blank at the same time.
+
+* `network_ids` - (Required) Specifies the subnet Network ID.
+
+* `name` - (Optional) Human-readable name for the LoadBalancer.
```
Comment (anton-sidelnikov) [inline docs/resources/lb_loadbalancer_v3.md]: can be inhuman?)

**Review by anton-sidelnikov — CHANGES_REQUESTED** [review]

**Review by outcatcher — APPROVED** [review]

**Review by anton-sidelnikov — APPROVED** [review]

**Pull request merged**